### ISSUE TITLE
AOSC OS MIPS64r6el Stage 3 Progress

### ISSUE THREAD

**Issue by MingcongBai** [AOSC-CIP-Pilot/aosc-os-abbs#2]
This issue tracks the tasks completed/pending for a Stage 3 (full-featured and production-ready) AOSC OS Base, Container, BuildKit, and Server releases. Upon completion, we will have a range of usable AOSC OS distributions for CIP's development needs.

The completion of Stage 3 also marks the beginning of upstream inter-operation with AOSC-Dev. Toolchain and packaging changes will begin to enter upstream. This port will also enter maintenance mode, with the following goals in mind:

- Tracking updates from both AOSC and software to keep this port up-to-date.
- Introduce all packages from the tree, where applicable (and mark all `FAIL_ARCH`).
- Preparation of training materials to get new interns ready for distribution maintenance and CI/CD introduction.
- Report any and all issues encountered during build time to upstream.

### Road to Stage 3

- [x] Rebuilding all non-`noarch` packages from Stage 2 with full feature and dependencies using Ciel
- [x] Completing distribution recipes
    - [x] Step 1: Container (non-bootable)
    - [x] Step 2: Base (bootable)
    - [x] Step 3: BuildKit (bootable)
    - [x] Step 4: Server (bootable)
- [x] Open pull request to AOSC-Dev upstream
    - [x] aosc-os-abbs - pull requests: [#4294 Core 10 (merged)](https://github.com/AOSC-Dev/aosc-os-abbs/pull/4294), pull request [merged](https://github.com/AOSC-Dev/aosc-os-abbs/pull/4351) for `cip/mips64r6el-bring-up`.
    - [x] autobuild3 - pull requests: [#138 (merged)](https://github.com/AOSC-Dev/autobuild3/pull/138), [#139 (merged)](https://github.com/AOSC-Dev/autobuild3/pull/139)
    - [x] acbs - merged upstream
- [ ] Report any and all issues encountered during build time to upstream.
- [ ] Update and translate Autobuild3/ACBS documentations to reflect latest changes

### Step 0: Rebuilding

Rebuild all non-`noarch` packages from Stage 2. The build list below is generated using the following command in a Stage 2 snapshot.

```bash
apt list | grep /rolling | grep mips64r6el | cut -f1 -d/ | sed -e '/-dbg$/d' | sed -e 's|^|- [ ] |g'
```

Note that the list below does not need to be built in order. Make sure that the `ABSTAGE2=1` flag is removed and each package is built using Ciel. 

#### Toolchain Adjustment

Rebuild Core and adjust Autobuild3

- [x] GCC: introduce three patches ([1](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=354c97ee05b23707efbfc7ff6640dcec7336ac7a), [2](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=4479f1dc79fc4f1b5e0fed209df35f405bc94589), [3](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=593632051f48a20bdc685d00d168f064d808bd7b)) - [committed to aosc-os-abbs:cip/mips64r6el-bring-up](https://github.com/AOSC-CIP-Pilot/aosc-os-abbs/commit/e4084cb490aa4e0c0ff194eebf7e25bcba79827a)
- [x] Add `-mcompact-branches=always` to default flags - [committed to autobuild3:cip/mips64r6el](https://github.com/AOSC-CIP-Pilot/autobuild3/commit/8892c0547f53ec56634da5f1a59bb06032a453c0)
- [x] Glibc: Patch a [known issue](https://sourceware.org/bugzilla/show_bug.cgi?id=29730) with >= 2.35
- [x] Binutils: Patch an [issue where the default output format does not conform with target specifications](https://sourceware.org/bugzilla/show_bug.cgi?id=25494)
- [x] Rebuild `groups/core`

#### Versions to Calibrate

Some packages were either wrongfully updated, or were missing REL bumps. To be corrected in Stage 3. Rebuild these packages first and check each box upon calibration.

Also take note to archive each package from the [repository](https://repo.oss.cipunited.com/aosc-mipsr6/debs/) upon completion, in order to avoid confusion.

- [x] arch-install-scripts, 24 => 24-1
- [x] bind, 9.18.8 => 9.16.21-1
- [x] cgdb, 0.8.0 => 0.7.1-2
- [x] expat, 2.4.9 => 2.4.6
- [x] iotop, 0.6-3 => 0.6-4
- [x] mdadm, 4.1-5 => 4.1-4
- [x] ~mtd-utils, 2.1.5 => 2.1.3~ - 2.1.5 is a necessary update to fix build errors against newer Linux API headers.
- [x] pango, 1.50.11 => 1.50.8
- [x] polkit, 121 => 0.120
- [x] ~~postfix, 3.7.3 => 3.4.8-4~~ - An update is necessary to fix FTBFS on Linux 6.x, even with the patch. Recablirating mariadb and rebuild.
- [x] python-3, 3.10.8 => 3.10.0
- [x] quota-tools, 4.06 => 4.06-1
- [x] sphinx, 5.3.0 => 4.2.0
- [x] xfsprogs, 5.14.2 => 5.14.2

#### To Rebuild

Then, rebuild all packages listed below, check upon completion.

- [x] acbs
- [x] acl
- [x] admin-base
- [x] aosc-aaa
- [x] apt
- [x] arp-scan
- [x] aspell
- [x] attr
- [x] babeltrace
- [x] bash
- [x] bc
- [x] bcache-tools
- [x] bind
- [x] binutils
- [x] bison
- [x] bluez
- [x] boost
- [x] btrfs-progs
- [x] bzip2
- [x] c-ares
- [x] cairo
- [x] cffi
- [x] cgdb
- [x] check
- [x] cmake - Blocked by emacs (which brings everything else)
- [x] core-base
- [x] coreutils
- [x] cpio
- [x] cracklib
- [x] cryptsetup
- [x] ctags
- [x] curl
- [x] cyrus-sasl
- [x] db
- [x] dbus-glib
- [x] dbus
- [x] debianutils
- [x] debug-base
- [x] dhcp
- [x] diffutils
- [x] ding-libs
- [x] dnsmasq
- [x] dos2unix
- [x] dosfstools
- [x] doxygen
- [x] dpkg
- [x] duktape
- [x] e2fsprogs
- [x] ed
- [x] elfutils
- [x] ethtool
- [x] exfat-utils
- [x] expat
- [x] expect
- [x] f2fs-tools
- [x] fcron
- [x] file
- [x] findutils
- [x] flex
- [x] fontconfig
- [x] freeglut
- [x] freetype
- [x] fribidi
- [x] fuse
- [x] gawk
- [x] gc
- [x] gcc-runtime
- [x] gcc
- [x] gdb
- [x] gdbm
- [x] geoip-api-c
- [x] gettext
- [x] ghostscript - dependency hell: cups
- [x] giflib
- [x] git - dependency hell: cups
- [x] glib
- [x] glibc
- [x] glslang
- [x] glu
- [x] gmp
- [x] gnupg
- [x] gnutls
- [x] gobject-introspection
- [x] gpart
- [x] gperf
- [x] gpgme - dependency hell: qt-5
- [x] gpm
- [x] gptfdisk
- [x] graphite
- [x] grep
- [x] groff
- [x] gssproxy
- [x] guile
- [x] gzip
- [x] harfbuzz
- [x] hddtemp
- [x] hdparm
- [x] hfsprogs
- [x] htop
- [x] http-parser
- [x] hwloc
- [x] icu
- [x] idnkit
- [x] ijs
- [x] imlib2
- [x] inetutils
- [x] inih
- [x] iproute2
- [x] iptables
- [x] isl
- [x] iw
- [x] iwd
- [x] jansson
- [x] jasper
- [x] jemalloc
- [x] jfsutils
- [x] joe
- [x] json-c
- [x] kbd
- [x] kernel-tools
- [x] keyutils
- [x] kmod
- [x] krb5
- [x] lcms2
- [x] less
- [x] libaio
- [x] libapt-pkg-perl
- [x] libarchive
- [x] libassuan
- [x] libatomic-ops
- [x] libbloom
- [x] libbsd
- [x] libcap-ng
- [x] libcap
- [x] libcork - Sphinx: failed to build documentation.
- [x] libdaemon
- [x] libdatrie
- [x] libdrm
- [x] libedit
- [x] libev
- [x] libevent
- [x] libffi
- [x] libgcrypt
- [x] libgee
- [x] libglvnd
- [x] libgpg-error
- [x] libgudev
- [x] libical
- [x] libid3tag
- [x] libidn2
- [x] libidn
- [x] libipset
- [x] libjpeg-turbo
- [x] libksba
- [x] libmaxminddb
- [x] libmbim
- [x] libmicrohttpd
- [x] libmnl
- [x] libndp
- [x] libnftnl
- [x] libnl
- [x] libnsl2
- [x] libpaper
- [x] libpcap
- [x] libpipeline
- [x] libpng
- [x] libpsl
- [x] libpwquality
- [x] libqmi
- [x] libqrtr-glib
- [x] libseccomp
- [x] libsigsegv
- [x] libsodium - `AUTOTOOLS_AFTER__MIPS64R6EL="${AUTOTOOLS_AFTER__NONX86}"`
- [x] libssh
- [x] libtasn1
- [x] libteam
- [x] libthai
- [x] libtiff
- [x] libtirpc
- [x] libtool
- [x] libunistring
- [x] libunwind
- [x] liburcu
- [x] libusb
- [x] libutempter
- [x] libuv
- [x] libwebp
- [x] libxau
- [x] libxcb
- [x] libxdmcp
- [x] libxml2
- [x] libxslt
- [x] linux+api
- [x] linux-kernel-6.0.0 - Pending Kernel config review (for Boston and "virt" platforms).
- [x] linux-pam
- [x] lksctp-tools
- [x] llvm-runtime
- [x] llvm
- [x] lm-sensors
- [x] lmdb
- [x] lrzip
- [x] lshw
- [x] lsof
- [x] lua
- [x] lvm2
- [x] lz4
- [x] lzip
- [x] lzo
- [x] lzop
- [x] m4
- [x] make
- [x] man-db
- [x] markupsafe
- [x] mbedtls
- [x] mdadm
- [x] mercurial
- [x] mesa
- [x] modemmanager
- [x] mpc
- [x] mpdecimal
- [x] mpfr
- [x] mtd-utils
- [x] mtr
- [x] multipath-tools
- [x] nano
- [x] ncurses
- [x] net-tools
- [x] netcat
- [x] nethogs
- [x] nettle
- [x] network-base
- [x] networkmanager (Still, with LTO turned off)
- [x] newt
- [x] nfs-utils
- [x] nghttp2
- [x] ninja
- [x] npth
- [x] nspr
- [x] nss
- [x] numactl
- [x] nvme-cli
- [x] ocaml
- [x] ofono
- [x] openldap
- [x] openmpi
- [x] openssh
- [x] openssl
- [x] os-prober
- [x] p11-kit
- [x] p7zip - Why does it require wxgtk3?
- [x] pacman - dependency hell: gpgme => qt-5
- [x] pahole
- [x] pango
- [x] parted
- [x] patch
- [x] pciutils
- [x] pcre2
- [x] pcre
- [x] pcsclite
- [x] perl-html-parser
- [x] perl-locale-gettext
- [x] perl-net-ssleay
- [x] perl-text-charwidth
- [x] perl-unicode-linebreak
- [x] perl-xml-parser
- [x] perl
- [x] pixman
- [x] pkg-config
- [x] polkit
- [x] popt
- [x] postfix
- [x] powertop
- [x] ppp
- [x] procps
- [x] progress
- [x] proxychains-ng
- [x] psmisc
- [x] pv
- [x] pycairo
- [x] pycryptodome
- [x] pygobject-3
- [x] python-2
- [x] python-3
- [x] python-base
- [x] quota-tools
- [x] ragel
- [x] rarian
- [x] re2c
- [x] readline
- [x] reiserfsprogs
- [x] rfkill
- [x] rpcbind
- [x] rpcsvc-proto
- [x] rrdtool
- [x] rsync
- [x] screen
- [x] sdparm
- [x] sed
- [x] sg3-utils
- [x] shadow
- [x] shadowsocks-libev
- [x] slang
- [x] smartmontools
- [x] source-highlight
- [x] spdylay
- [x] sqlite
- [x] strace
- [x] sudo
- [x] swig
- [x] sysfsutils
- [x] systemd
- [x] systemtap
- [x] tar
- [x] tcl
- [x] texinfo
- [x] thin-provisioning-tools
- [x] tinycdb
- [x] tk
- [x] tmux
- [x] tree
- [x] trousers
- [x] uboot-tools
- [x] udns
- [x] unrar
- [x] unzip
- [x] update-alternatives
- [x] usbutils
- [x] util-base
- [x] util-linux
- [x] vala
- [x] valgrind
- [x] vim - Dependency hell: gtk-3
- [x] w3m
- [x] wget
- [x] which
- [x] wireless-tools
- [x] wpa-supplicant
- [x] x11-app
- [x] x11-font
- [x] x11-lib
- [x] xcb-util
- [x] xfsprogs
- [x] xmlto
- [x] xxhash
- [x] xz
- [x] yaml
- [x] zip
- [x] zlib
- [x] zstd

</details>

### Step 1: Container

Recipe items to complete for Container.

- [x] admin-base
- [x] core-base
- [x] editor-base
- [x] network-base
- [x] python-base
- [x] systemd-base
- [x] util-base
- [x] web-base

### Step 2: Base

Recipe items to complete for Base.

- [x] aosc-os-presets-base
- [x] audio-base
- [x] boot-base
- [x] kernel-base
- [x] kmscon-base

### Step 3: BuildKit

Recipe items to complete for BuildKit.

- [x] acbs
- [x] autobuild3
- [x] debug-base
- [x] devel-base
- [x] git

### Step 4: Server

Recipe items to complete for Server.

- [x] server-base

*Issues, problem, and errors encountered during this build will be recorded in the comments below.*

**Comment by MingcongBai** [comment]
Please refer to a log of known issues from #1. Stage 3 may or may not address them.

### Unmet Essential Dependencies
The following lists a collection of devel packages which does not work or are just unsupported on this platform.

- Rust (can be cross-compiled, produces SIGSEGV during native build, but cross-compiled toolchain can compile usable simple Rust program - 1.60, glibc 2.31)
- Go (no architecture support)
- Node.js (V8 is known not to work on this architecture, according to @FlyGoat)

**Comment by MingcongBai** [comment]
Whilst trying to start an `mips64r6el` `systemd-nspawn` container, systemd would freeze upon intialisation. Upon investigation, we have found one workaround and one finalised (and upstream) fix, thanks to tips from @Icenowy and @FlyGoat.

- Workaround: Drop `-fira-loop-pressure` from CFLAGS may resolve this issue.
- Fix: Apply [this patch](https://lists.gnu.org/archive/html/qemu-devel/2022-10/msg05496.html) by @Icenowy against Qemu 7.1.0. This patch should soon see upstreaming.

**Comment by MingcongBai** [comment]
Another note from @Cyanoxygen and @liushuyu. Without systemd support in `dbus`, Ciel containers will timeout during startup, this is because the host needs to initiate a D-Bus connection with the container. @Cyanoxygen has sine introduced comments and instructions on how to handle the `dbus` package in stage 2.

See `dbus`'s [autobuild/defines](https://github.com/AOSC-CIP-Pilot/aosc-os-abbs/blob/7f1595bfa952a67b344c02195ee93ca1f36c36ee/extra-admin/dbus/autobuild/defines.stage2).

**Comment by MingcongBai** [comment]
To standardise and record stage 2 procedures, I have [implemented](https://github.com/AOSC-CIP-Pilot/autobuild3/commits/cip/stage2-suffix) `.stage2`-suffix support (enabled via `ABSTAGE2=1`) in Autobuild3, whereas @Cyanoxygen [implemented](https://github.com/AOSC-CIP-Pilot/acbs/commit/67d2bc7c3f45d2473073aa641207405d65351930) `ABSTAGE2` switch detection in ACBS.

These changes will be upstreamed at AOSC-Dev upon successful testing.

### Upstreamed Changes

- [x] ACBS
- [x] Autobuild3

**Comment by MingcongBai** [comment]
### Additional .stage2 Packages

More packages needed dependency bootstrapping. Rebuild these packages after stage 3 rebuild.

- [x] colord
- [x] cups
- [x] emacs
- [x] gstreamer
- [x] libgd
- [x] libnice
- [x] pinentry
- [x] poppler
- [x] pulseaudio
- [x] qt-4
- [x] qt-5
- [x] subversion
- [x] wayland
- [x] wheel

**Comment by MingcongBai** [comment]
To patch: Known issue with Glibc >= 2.35.

[29730 - broken y2038 support in fstatat on MIPS N64](https://sourceware.org/bugzilla/show_bug.cgi?id=29730)

**Comment by Cyanoxygen** [comment]
LLVM/Clang should be patched to “recongnize” this architecture because the default triplet is “unknown”.

- [x] Patch LLVM/Clang

Edit: Clang is blocking packages.
Edit: LLVM rebuild is blocked by findlib, which is unable to link against OCaml libraries.
Edit: Ocaml is being rebuilt with LTO disabled.

Update: LLVM is being rebuilt.

Update: LLVM is now able to compile programs, without specifying target triple. Thanks!

**Comment by liushuyu** [comment]
> LLVM/Clang should be patched to “recongnize” this architecture because the default triplet is “unknown”.
> 
>     * [ ]  Patch LLVM/Clang
> 
> 
> Edit: Clang is blocking packages.

Hint: https://github.com/AOSC-CIP-Pilot/aosc-os-abbs/blob/43cbcbb21585279c0c1aa3360452133c77e49da2/base-devel/llvm/01-runtime/patches/0002-aosc-exclusive-tuples.patch.deferred#L42

**Comment by MingcongBai** [comment]
Issue when building GLib's testsuite...

```
[1/5] Linking target gio/tests/resources
FAILED: gio/tests/resources 
gcc  -o gio/tests/resources gio/tests/test_resources2.o gio/tests/resources.p/meson-generated_.._test_resources.c.o gio/tests/resources.p/meson-generated_.._test_resources2.c.o gio/tests/resources.p/meson-generated_.._digit_test_resources.c.o gio/tests/resources.p/meson-generated_.._test_resources_binary.c.o gio/tests/resources.p/resources.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,-O1,--sort-common,--as-needed -Wl,-build-id=sha1 -Wl,-z,relro -Wl,-z,now -specs=/usr/lib/gcc/specs/hardened-ld -flto -fuse-linker-plugin -pipe -Wno-error -fstack-protector-strong --param=ssp-buffer-size=4 -fexceptions -ggdb -O2 -march=mips64r6 -mtune=mips64r6 -mcompact-branches=always -fira-loop-pressure -fira-hoist-pressure -ftree-vectorize -specs=/usr/lib/gcc/specs/hardened-cc1 -flto=auto -D_GLIBCXX_ASSERTIONS -D_FORTIFY_SOURCE=2 '-Wl,-rpath,$ORIGIN/../../glib:$ORIGIN/../../gmodule:$ORIGIN/../../gobject:$ORIGIN/..' -Wl,-rpath-link,/var/cache/acbs/build/acbs.c5831njg/glib-2.72.3/abbuild/glib -Wl,-rpath-link,/var/cache/acbs/build/acbs.c5831njg/glib-2.72.3/abbuild/gmodule -Wl,-rpath-link,/var/cache/acbs/build/acbs.c5831njg/glib-2.72.3/abbuild/gobject -Wl,-rpath-link,/var/cache/acbs/build/acbs.c5831njg/glib-2.72.3/abbuild/gio -Wl,--start-group glib/libglib-2.0.so.0.7200.3 gmodule/libgmodule-2.0.so.0.7200.3 gobject/libgobject-2.0.so.0.7200.3 gio/libgio-2.0.so.0.7200.3 -Wl,--end-group
/usr/bin/ld: gio/tests/test_resources2.o: warning: linking abicalls files with non-abicalls files
/usr/bin/ld: gio/tests/test_resources2.o: linking mips:4000 module with previous mips:isa64r6 modules
/usr/bin/ld: gio/tests/test_resources2.o: linking -mnan=legacy module with previous -mnan=2008 modules
/usr/bin/ld: failed to merge target specific data of file gio/tests/test_resources2.o
collect2: error: ld returned 1 exit status
```

Suspecting a failure to pass in necessary flags for some of the objects, as evidenced by the log above.

**Comment by Cyanoxygen** [comment]
According to upstream's  [#2768](https://gitlab.gnome.org/GNOME/glib/-/issues/2768) and [!2923](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2923), I think we better disable the test suite compilation.

**Comment by Cyanoxygen** [comment]
libwmf requires freetype2, but `ft2build.h` does not reside in libwmf's desired location (it is at `/usr/include/freetype2`), libwmf's config system tries to include this header file without `-I/usr/include/freetype2` .

**Comment by Cyanoxygen** [comment]
Every version updates/downgrades and architecture specific changes will be upstreamed to AOSC-Dev/aosc-os-abbs later. 

For a list of modified packages, please see `git diff` - I will fill out the reason for the change in the comments of the respective file, if I can recall them (remove the comments when commiting changes).

**Comment by Cyanoxygen** [comment]
- `hidapi` is now under https://github.com/libusb/hidapi . Without update it will fail during RECONF.
- `systemd` build failed, due to obsolete `markupsafe` package. Updating to the latest version.
  Edit: It looks like it is already the latest version.
  Edit: It is `jinja2` who needs an update :)
  Edit: jinja2 needs a rebuild. Its `spec` points to the latest version, but no actual package available.

**Comment by Cyanoxygen** [comment]
Use the `linux64-mips64` target for `openssl`.
- ~Rebuild OpenSSL~

---

```
gcc  -I. -Iinclude -fPIC -pthread -mabi=64 -mips3 -Wa,--noexecstack -pipe -Wno-error  -fstack-protector-strong --param=ssp-buffer-size=4  -fexceptions  -ggdb    -O3    -march=mips64r6 -mtune=mips
64r6 -mcompact-branches=always        -fira-loop-pressure -fira-hoist-pressure -ftree-vectorize  -specs=/usr/lib/gcc/specs/hardened-cc1  -flto=auto            -Wa,--noexecstack  -D_GLIBCXX_ASSERT
IONS  -D_FORTIFY_SOURCE=2                          -pipe -Wno-error  -fstack-protector-strong --param=ssp-buffer-size=4  -fexceptions  -ggdb    -O3    -march=mips64r6 -mtune=mips64r6 -mcompact-br
anches=always        -fira-loop-pressure -fira-hoist-pressure -ftree-vectorize  -specs=/usr/lib/gcc/specs/hardened-cc1  -flto=auto            -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BN_ASM
_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/usr/lib/engines-1.1\"" -DZLIB -DNDEBUG -D_GLIBCXX_ASSERTIONS  -D_FORTIFY_SOURCE=2 
                        -MMD -MF apps/app_rand.d.tmp -MT apps/app_rand.o -c -o apps/app_rand.o apps/app_rand.c
Assembler messages:                             
Error: -mips3 conflicts with the other architecture options, which imply -mips64r6
cc1: error: '-mips3' conflicts with the other architecture options, which specify a mips64r6 processor
```

Forfeiting plan.

**Comment by Cyanoxygen** [comment]
PulseAudio's stage2 recipe does not take alternatives.stage2. While building `pulseaudio`, we found that autobuild3's stage2 mode does not take everything else with .stage2 suffix. A change is needed to address this.

- [x] Modify autobuild3 for more complete stage 2 support

This feature has been [implemented and tested](9af3184d8b4e722e0def309766f8d353d0fef70b) with `pulseaudio`.

**Comment by MingcongBai** [comment]
Node.js SIGILLs during build-time, despite setting `--with-mips-arch-variant=r6` in `configure`. @FlyGoat noted that his is a known issue, upstream issue pending (?).

```
root@mips64r6el-d15f1225 [ acbs.hxgl9qz1 ] ? LD_LIBRARY_PATH=/var/cache/acbs/build/acbs.hxgl9qz1/nodejs/out/Release/lib.host:/var/cache/acbs/build/acbs.hxgl9qz1/nodejs/out/Release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../.; mkdir -p /var/cache/acbs/build/acbs.hxgl9qz1/nodejs/out/Release/obj/gen; "/var/cache/acbs/build/acbs.hxgl9qz1/nodejs/out/Release/node_mksnapshot" "/var/cache/acbs/build/acbs.hxgl9qz1/nodejs/out/Release/obj/gen/node_snpshot.cc"                                                                                       
qemu: uncaught target signal 4 (Illegal instruction) - core dumped
Illegal instruction (core dumped)
```

Adjusting recipes accordingly to avoid `nodejs` dependency.

**Comment by MingcongBai** [comment]
Wayland fails to generate documentation using Graphviz's `dot` command, due to a SIGILL exception:

```
[6/107] Generating doc/doxygen/xml/wayland-architecture.png with a custom command
FAILED: doc/doxygen/xml/wayland-architecture.png 
/usr/bin/dot -Tpng -odoc/doxygen/xml/wayland-architecture.png ../doc/doxygen/dot/wayland-architecture.gv
qemu: uncaught target signal 4 (Illegal instruction) - core dumped
```

Disabling documentation for now, pending investigation.

**Comment by MingcongBai** [comment]
Current system is built without MSA, as it is optional for the P6600 core design. CIP is actively evaluating whether this core (which likely won't see commercial use) should be considered for support. If MSA were to be enabled, we will rebuild the whole repository.

Source: @FlyGoat

**Comment by MingcongBai** [comment]
A potential proposal to remove the `gdm` build dependency from `pulseaudio`, use tmpfiles.d's `a` marker for setting POSIX ACL attributes.

[Reference](https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html).

**Comment by Cyanoxygen** [comment]
Glibc 2.36, which Core 10 planned to use, has introduced some changes to its header files, one particular of them is `sys/mount.h`. This header file has a regression in its compatibility with Linux's `linux/mount.h`, which caused a bunch of redefinitions of macros and enums.

Upstream's release branch has introduced a set of patches to address this, see [this diff](https://sourceware.org/git/?p=glibc.git;a=blobdiff;f=sysdeps/unix/sysv/linux/sys/mount.h;h=19841d07385244813e3773d872348d37dcc8c97d;hp=df6b0dbb425fbf2e441696ce680369c4448125ea;hb=refs/heads/release/2.36/master;hpb=4dad97e2a2e510c6b53a0add29a2188714fcf4ab).

For more information, please see [this glibc wiki page](https://sourceware.org/glibc/wiki/Release/2.36#x86_64_.2864-bit.2C_Linux.29) and [this linuxquestions.org post](https://www.linuxquestions.org/questions/slackware-14/sysvinit-incompatibility-linux-mount-h-and-sys-mount-h-from-glibc-2-36-slackware-current-4175715474/).

Applying this patch should fix the header incompatibility error during various builds. This should be discussed during PR to https://github.com/AOSC-Dev/aosc-os-abbs.

**Comment by Cyanoxygen** [comment]
`archlinux-keyring` is unable to build on mips64r6el, because Rust is unavailable, using package from mainline repository.

Edit: All noarch packages now officially being stolen from mainline repository, skipping noarch packages rebuild.

**Comment by MingcongBai** [comment]
### Additional Version Calibrations

We have neglected to bump REL on some of the packages during Stage 3 to adapt them for current system dependencies and/or the mips64r6el architecture. Rebuild the following package to close out the rebuild process.

- [x] accountsservice
- [x] aosc-artwork
- [x] autofs
- [x] cifs-utils
- [x] desktop-base
- [x] findlib
- [x] glib-networking
- [x] jack
- [x] libcaca
- [x] libclc
- [x] libotf
- [x] libsoup
- [x] libtasn1
- [x] libvdpau
- [x] libwmf
- [x] minizip
- [x] qt-6
- [x] sip

**Comment by MingcongBai** [comment]
We messed up. As we attempted to calibrate versions in step 0, the following packages' downgrade was overridden by pre-existing packages already uploaded to the CIP repository server. These changes will be merged back to AOSC-Dev, potentially causing merging problems with other topics.

```
cgdb/now 0.8.0 mips64r6el [installed,local]
exfat-utils/now 1.3.0-1 mips64r6el [installed,local]
expat/now 2.4.9 mips64r6el [installed,local]
libpng/now 1.6.37-2 mips64r6el [installed,local]
mdadm/now 4.1-5 mips64r6el [installed,local]
openssh/now 8.8p1 mips64r6el [installed,local]
polkit/now 1:121 mips64r6el [installed,local]
python-3/now 3.10.8-1 mips64r6el [installed,local]
xfsprogs/now 5.14.2-1 mips64r6el [installed,local]
```

**Comment by Cyanoxygen** [comment]
To avoid this in the future, first, we need to add another checklist to make sure that every calibrated package is removed from the stage 2 repository, and, try to avoid package updates during stage 2. In the case of an FTBFS package, mark it for updating and recalibration.

I apologize for this mess, I forgot to remove the newer packages from the stage 2 repository.

**Comment by MingcongBai** [comment]
As this port is now merged as a secondary architecture in mainline, we will proceed with routine updates and fixes at the [upstream repository](https://github.com/AOSC-Dev/aosc-os-abbs). This repository will be used for bleeding-edge updates (toolchains and kernels, etc.).

We are currently rebuilding the full R6 system release against the mainline repository.

**Comment by Cyanoxygen** [comment]
The beforementioned glibc's mount.h header regression is being fixed with Core 10.0.1.
